pytd/query_engine.py

```diff
@@ -126,43 +126,88 @@ class PrestoQueryEngine(QueryEngine):
 
     def __init__(self, apikey, endpoint, database, header):
         super(PrestoQueryEngine, self).__init__(apikey, endpoint, database, header)
-        self.engine = self._connect()
+        self.prestodb_connection, self.tdclient_connection = self._connect()
 
     @property
     def user_agent(self):
         """User agent passed to a Presto connection.
         """
-        return "pytd/{0} (prestodb/{1})".format(__version__, prestodb.__version__)
+        return "pytd/{0} (prestodb/{1}; tdclient/{2})".format(
+            __version__, prestodb.__version__, tdclient.__version__
+        )
 
-    def cursor(self):
+    @property
+    def presto_api_host(self):
+        """Presto API host obtained from ``TD_PRESTO_API`` env variable or
+        inferred from Treasure Data REST API endpoint.
+        """
+        return os.getenv(
+            "TD_PRESTO_API", urlparse(self.endpoint).netloc.replace("api", "api-presto")
+        )
+
+    def cursor(self, **kwargs):
         """Get cursor defined by DB-API.
 
         Returns
         -------
-        prestodb.dbapi.Cursor
+        prestodb.dbapi.Cursor, or tdclient.cursor.Cursor
         """
-        return self.engine.cursor()
+        if len(kwargs) == 0:
+            return self.prestodb_connection.cursor()
+
+        logger.warning(
+            "returning `tdclient.cursor.Cursor`. This cursor, `Cursor#fetchone` "
+            "in particular, might behave different from your expectation, "
+            "because it actually executes a job on Treasure Data and fetches all "
+            "records at once from the job result."
+        )
+
+        original_cursor_kwargs = self.tdclient_connection._cursor_kwargs.copy()
+
+        params = self.tdclient_connection._cursor_kwargs
+        if "type" in kwargs:
+            params["type"] = kwargs["type"]
+        if "db" in kwargs:
+            params["db"] = kwargs["db"]
+        if "result_url" in kwargs:
+            params["result_url"] = kwargs["result_url"]
+        if "priority" in kwargs:
+            params["priority"] = kwargs["priority"]
+        if "retry_limit" in kwargs:
+            params["retry_limit"] = kwargs["retry_limit"]
+        if "wait_interval" in kwargs:
+            params["wait_interval"] = kwargs["wait_interval"]
+        if "wait_callback" in kwargs:
+            params["wait_callback"] = kwargs["wait_callback"]
```

Comment:
It'd be better to have a warning for unknown `kwargs` since it tends to bypass typo.

Reply:
7abbabbb1b174dd0ff1104294292bbf12f22c174